File: src/browser/git/file.ts
import * as fs from "fs-extra";
import * as path from "path";
import * as chardet from "chardet";
import * as iconv from "iconv-lite";
import { exec } from "./exec";

export async function catFile(repository: string, file: FileSpec): Promise<Buffer> {
  if (file.revspec === "UNSTAGED") {
    const abspath = path.join(repository, file.path);
    return fs.readFile(abspath);
  }
  const target = `${file.revspec === "STAGED" ? "" : file.revspec}:${file.path}`;
  const { stdout } = await exec("show", {
    repository,
    args: ["-p", target]
  });
  return stdout;
}

export async function getTextFileContent(repository: string, file: FileSpec): Promise<TextFile> {
  const buffer = await catFile(repository, file);
  const encoding = chardet.detect(buffer) || "utf8";
  const content = iconv.decode(buffer, encoding);
  return { ...file, encoding, content };
}

export async function saveTo(repository: string, file: FileSpec, destPath: string): Promise<void> {
  const stdout = await catFile(repository, file);
  await fs.writeFile(destPath, stdout);
}

File: src/rust-backend/native/src/git/file.rs
use super::{exec, GitError};
use std::fs::File;
use std::io::prelude::Write;
use std::path::Path;

pub fn get_content(repo_path: &Path, rel_path: &str, revspec: &str) -> Result<Vec<u8>, GitError> {
    let target = if revspec == "STAGED" {
        format!(":{}", rel_path)
    } else {
        format!("{}:{}", revspec, rel_path)
    };
    let args = vec!["-p", target.as_str()];
    let output = exec(repo_path, "show", &args, &[])?;
    GitError::assert_process_output("show", &output)?;

    Ok(output.stdout)
}

pub fn save_to(
    repo_path: &Path,
    rel_path: &str,
    revspec: &str,
    dest_path: &str,
) -> Result<(), GitError> {
    let content = get_content(repo_path, rel_path, revspec)?;
    let mut file = File::create(dest_path)?;
    file.write_all(&content)?;
    Ok(())
}
